core/jvm/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2
+package interop
+package flow
+
+import cats.MonadThrow
+import cats.effect.kernel.Async
+import cats.syntax.all._
+
+import java.util.concurrent.Flow.{Subscriber, Subscription}
+import java.util.concurrent.atomic.AtomicReference
+
+/** Implementation of a [[Subscriber]].
+  *
+  * This is used to obtain a [[Stream]] from an upstream reactive-streams system.
+  *
+  * @see [[https://github.com/reactive-streams/reactive-streams-jvm#2-subscriber-code]]
+  */
+private[flow] final class StreamSubscriber[F[_], A](
+    private[flow] val subscriber: StreamSubscriber.FSM[F, A]
+)(implicit
+    F: MonadThrow[F]
+) extends Subscriber[A] {
+
+  /** Called by an upstream reactive-streams system. */
+  override def onSubscribe(subscription: Subscription): Unit = {
+    nonNull(subscription)
+    subscriber.onSubscribe(subscription)
+  }
+
+  /** Called by an upstream reactive-streams system. */
+  override def onNext(a: A): Unit = {
+    nonNull(a)
+    subscriber.onNext(a)
+  }
+
+  /** Called by an upstream reactive-streams system. */
+  override def onComplete(): Unit =
+    subscriber.onComplete()
+
+  /** Called by an upstream reactive-streams system. */
+  override def onError(t: Throwable): Unit = {
+    nonNull(t)
+    subscriber.onError(t)
+  }
+
+  /** Creates a [[Stream]] from this [[Subscriber]]. */
+  def stream(subscribe: F[Unit]): Stream[F, A] =
+    subscriber.stream(subscribe)
+}
+
+private[flow] object StreamSubscriber {
+
+  /** Instantiates a new [[StreamSubscriber]] for the given buffer size. */
+  def apply[F[_], A](bufferSize: Int)(implicit F: Async[F]): F[StreamSubscriber[F, A]] =
+    fsm[F, A](bufferSize).map(fsm => new StreamSubscriber(subscriber = fsm))
```

Comment:
Should we throw an exception if `bufferSize` is not positive?

Reply:
Sure, I think Cats Effect does this in `std` with `require`. So this value can't be zero?

Reply:
Luis explained offline that the `bufferSize` is used for `request(...)`, which is required positive by the spec.